File: training-data/kotlin/LruCache.kt
class LruCache<K, V>(private val capacity: Int) {
    private val map = LinkedHashMap<K, V>(capacity, 0.75f, true)

    fun get(key: K): V? = map[key]

    fun put(key: K, value: V) {
        if (map.size >= capacity && !map.containsKey(key)) {
            val oldest = map.keys.first()
            map.remove(oldest)
        }
        map[key] = value
    }
}

fun main() {
    val cache = LruCache<Int, Int>(2)
    cache.put(1, 1)
    cache.put(2, 2)
    println(cache.get(1))
    cache.put(3, 3)
    println(cache.get(2))
}

File: training-data/rust/lru_cache.rs
use std::collections::HashMap;

struct LruCache {
    capacity: usize,
    order: Vec<i32>,
    map: HashMap<i32, i32>,
}

impl LruCache {
    fn new(capacity: usize) -> Self {
        LruCache { capacity, order: Vec::new(), map: HashMap::new() }
    }

    fn touch(&mut self, key: i32) {
        self.order.retain(|&k| k != key);
        self.order.push(key);
    }

    fn get(&mut self, key: i32) -> Option<i32> {
        if let Some(&value) = self.map.get(&key) {
            self.touch(key);
            Some(value)
        } else {
            None
        }
    }

    fn put(&mut self, key: i32, value: i32) {
        if self.map.len() >= self.capacity && !self.map.contains_key(&key) {
            let lru_key = self.order.remove(0);
            self.map.remove(&lru_key);
        }
        self.map.insert(key, value);
        self.touch(key);
    }
}

fn main() {
    let mut cache = LruCache::new(2);
    cache.put(1, 100);
    cache.put(2, 200);
    cache.get(1);
    cache.put(3, 300);
    println!("{:?} {:?} {:?}", cache.get(1), cache.get(2), cache.get(3));
}
